src/main/java/duke/task/TaskList.java

```diff
@@ -0,0 +1,405 @@
+package duke.task;
+
+import java.util.ArrayList;
+
+import duke.exception.DukeCommandDoesNotExistException;
+import duke.exception.DukeException;
+import duke.exception.DukeNoDescriptionException;
+import duke.exception.DukeNoTimeSpecifiedException;
+import duke.exception.DukeOutOfBoundException;
+import duke.io.UserInput;
+import duke.ui.Ui;
+
+/**
+ * This is a TaskList class that used to store the tasks in an ArrayList.
+ */
+public class TaskList {
+    private final ArrayList<Task> list = new ArrayList<>(); // arraylist to keep track of task
+
+    /**
+     * List out all the tasks currently in the task list.
+     *
+     * @param userInput A UserInput object, capturing the user's input.
+     * @throws DukeException If user's input is invalid.
+     */
+    public void listTask(UserInput userInput) throws DukeException {
+        if (!userInput.getDescription().equals("") || !userInput.getTime().equals("")) {
+            // throw wrong command exception
+            throw new DukeCommandDoesNotExistException("OOPS!!! This command does not exist.");
+        }
+        Ui.drawDivider();
+        if (list.size() == 0) {
+            System.out.println("Nothing on your list!");
+        } else {
+            System.out.println("Here are the tasks in your list:");
+            for (int i = 0; i < list.size(); i++) {
+                System.out.printf("%d.%s%n", i + 1, list.get(i).toString());
+            }
+        }
+        System.out.printf("Now you have %d tasks in the list%n", list.size());
+        Ui.drawDivider();
+    }
+
+    /**
+     * Add task to the task list based on the user's input.
+     * Print a message if the task is successfully added to the task list.
+     *
+     * @param userInput A UserInput object, capturing the user's input.
+     * @throws DukeException If user's input is invalid.
+     */
+    public void addTask(UserInput userInput) throws DukeException {
+        // add ToDoTask
+        if (userInput.getCommand().equals("todo")) {
+            if (userInput.getDescription().equals("")) {
+                // throw no description exception
+                throw new DukeNoDescriptionException("OOPS!!! The description of a todo cannot be empty.");
+            }
+            Task newTask = new ToDoTask(userInput.getDescription());
+            list.add(newTask);
+            Ui.drawDivider();
+            System.out.println("Got it. I've added this task:");
+            System.out.println(newTask);
+        }
+
+        // add DeadlineTask
+        if (userInput.getCommand().equals("deadline")) {
+            if (userInput.getDescription().equals("")) {
+                // throw no description exception
+                throw new DukeNoDescriptionException("OOPS!!! The description of a deadline cannot be empty.");
```

Comment:
You could consider handling these exception in the constructor of the relevent tasks. i.e throws these exception while creating the task during ->
 ```Java 
new Deadline("{empty description}", "{incorrect time format}");
```

Reply:
Alright, thanks for the recommendation, will think about it.